docs/OAUTH2.md

```diff
@@ -0,0 +1,208 @@
+- `GET /oauth/client/{client id}`
+- `POST /oauth/client`
+
+  ```json
+  {
+    "clientName": "Hometown SIS",
+    "roles": [
+      "vendor"
+    ]
+  }
+  ```
+
+- `PUT /oauth/client/{client id}`
+
+  ```json
+  {
+    "client_id": "{client id}",
+    "clientName": "Hometown SIS",
+    "roles": [
+      "vendor"
+    ]
+  }
+  ```
+
+- `DELETE /oauth/client/{client id}` - deactivates, does not remove the client
+- `POST /oauth/client/{client id}/reset` - generates a new `client_secret` for
+  the key
+
+### Server Configuration
+
+These environment variables control the internal OAuth2 provider:
+
+| Variable                 | Default Value    | Purpose                                                |
+| ------------------------ | ---------------- | ------------------------------------------------------ |
+| OAUTH_EXPIRATION_MINUTES | 60               | Length of time for a token to remain valid, in minutes |
+| OAUTH_TOKEN_ISSUER       | ed-fi-meadowlark | Value for the `iss` (issuer) claim on the token        |
+| OAUTH_TOKEN_AUDIENCE     | ed-fi-meadowlark | Value for the `aud` (audience) claim on the token      |
+| OAUTH_SIGNING_KEY        | {empty}          | Base64-encoded 256 bit signing key                     |
+
+There are multiple options for generating a signing key.
+
+1. If you have `openssl`, run: `openssl rand -base64 256`.
+2. Run the Meadowlark service and issue a GET request to endpoint
+   `/{stage}/oauth/createSigningKey`
+
+## Meadowlark as OAuth2 Client
+
+As described above, Meadowlark's Ed-Fi API treats the OAuth2 provider as a
+service, rather than using shared code. Done right, this should make it trivial
+to integrate with third party providers.
+
+One implication: the API application should not need to know the signing key;
+only the OAuth2 provider will have that. Thus the client application must use
+the token introspection endpoint to validate a received token. Validated tokens
+are cached so that subsequent requests do not need to revalidate them, except
+for confirming the expiration date. Tokens remain in the cache by default for up
+to 5 minutes; if there are 1000 (configurable) tokens then the oldest ones are
+vacated more quickly to make room for more tokens.
+
+### Client Configuration
+
+These environment variables are on the Meadowlark client side; when we get to
+the point of having third party OAuth providers, these settings will still be
+relevant, whereas the settings above will not be needed.
+
+| Variable                                      | Default Value     | Purpose                                               |
+| --------------------------------------------- | ----------------- | ----------------------------------------------------- |
+| OAUTH_CLIENT_PROVIDED_TOKEN_CACHE_TTL         | 300000 ms (5 min) | Length of time to accept a token without revalidating |
+| OAUTH_CLIENT_PROVIDED_TOKEN_CACHE_MAX_ENTRIES | 1000              | Max number of validated tokens to cache               |
```

Comment:
What about the four env vars here:

https://github.com/Ed-Fi-Exchange-OSS/Meadowlark/blob/707bc47367a4f742a544a4b42372d458b68bd39f/Meadowlark-js/packages/meadowlark-core/src/middleware/OAuthFetch.ts#L12